USERMANUAL.md

```diff
@@ -0,0 +1,596 @@
+          from data
+          group by $column),
+        t2 as (
+          select $ref_column_id as col, count(*) as t2_n 
+          from `<dataset_id>.$ref_table_id`
+          group by $ref_column_id
+        )
+        select t1.*, t2.t2_n
+        from t1 
+        full outer join t2 using(col)
+        where not t1_n = t2_n 
+        
+rule_bindings:
+  EQUAL_COLUMNS:
+    entity_id: TEST_DATA
+    column_id: UNIQUE_KEY
+    row_filter_id: NONE
+    rule_ids:
+      - EQUAL_COLUMNS:
+          ref_table_id: <reference_table_id>
+          ref_column_id: <reference_column_id>
+```
+
+Full example: [Rule](docs/examples/row_by_row.yaml) and [rule binding](docs/examples/row_by_row.yaml)
+
+
+#### Equality of two columns (special case: columns do not contain repeated values)
+
+For this test we compare two columns that we know have distinct values (for example because we have checked this separately).
+
+```yaml
+rules:
+  EQUAL_COLUMNS_UNIQUE_VALUES:
+    rule_type: CUSTOM_SQL_STATEMENT
+    custom_sql_arguments:
+        - ref_table_id
+        - ref_column_id
+    params:
+      custom_sql_statement: |-
+        with t1 as (
+          select 
+            $column as t1_col, 
+            row_number() over (order by $column) as n 
+          from data),
+        t2 as (
+          select 
+            $ref_column_id as t2_col, 
+            row_number() over (order by $ref_column_id) as n 
+            from `<dataset_id>.$ref_table_id`
+        )
+        select t1.t1_col, t2.t2_col
+        from t1 
+        full outer join t2 using(n)
+        where not t1_col = t2_col 
+
+rule_bindings:
+  EQUAL_COLUMNS:
+    entity_id: TEST_DATA
+    column_id: UNIQUE_KEY
+    row_filter_id: NONE
+    rule_ids:
+      - EQUAL_COLUMNS_UNIQUE_VALUES:
+          ref_table_id: <reference_table_id>
+          ref_column_id: <reference_column_id>
+```
+
+Full example: [Rule](docs/examples/row_by_row.yaml) and [rule binding](docs/examples/row_by_row.yaml)
+
+
+#### Equality of an aggregate
+
+In this example we check that the aggregate in one table equals a value in another table. For instance, the sum over line items in an order should be equal to the total invoice amount in another table.
+
+This example uses a rule of type CUSTOM_SQL_STATEMENT with 5 parameters. We assume there are two tables, one where we will aggregated based on some key, and sum a value column, and another table that has the totals for each key. The parameters are 
+1. the column ID for the key and 
+1. the value that we are aggregated, 
+1. the column ID for the key and 
+1. the value in table that contains the totals, and 
+1. the table ID of the table containing the totals.
+
+As we specified everything that we need through the above “custom SQL arguments”, the column_id field in the rule binding has no effect. 
+
+```yaml
+rules:
+  EQUAL_AGGREGATE:
+    rule_type: CUSTOM_SQL_STATEMENT
+    custom_sql_arguments: 
+      key_column_id
```

Comment:
does this need to be a list? e.g. have a hyphen before each argument?

Reply:
No, it does not need that. I can see how for consistency/style you would want me to make it a list but I have tested it like this and it works. If I touch it, I would need to test it again. Wdyt..